### DISCUSSION TITLE
Why allow separate connections?

### DISCUSSION THREAD

**Discussion by wmertens** [enisdenjo/graphql-sse#46] [General]
It seems to me that all subscriptions should always come in over a single SSE channel that gets lazily opened.

So I wonder what the rationale is for making this optional?

**Comment by enisdenjo** [comment]
SSE is not multiplexed (like WebSockets), once the client connects - **only** the server can dispatch messages to the client.

The [GraphQL over Server-Sent Events Protocol](https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md) has 2 working modes:
- "single connection mode" which has the client establish a single SSE connection to the server and then use regular requests for operations.
- "distinct connection mode" which is the default behaviour, where the client passes the operation parameters along the SSE connection request and then the server streams the results.

Read more in the [GraphQL over Server-Sent Events Protocol](https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md).

**Comment by wmertens** [comment]
Hmm, but given that you can only have a few connections to a site allowed by the browser, shouldn't single connection mode be the only thing supported?

I'm trying to understand why this library supports distinct connection mode and defaults to it, it seems to me that it's technically inferior? Am I missing a benefit?

**Reply by enisdenjo** [reply]
HTTP/2 enabled servers can have much more connections, I think the browser default is 256 per domain. The need for "single connection mode" restriction comes from two reasons:

- HTTP/1 servers where browsers allow only 6 connections per domain
- A subscription heavy app communicating with HTTP/2 server where even the 256 connections is not enough
